server/pxf-api/src/main/java/org/greenplum/pxf/api/filter/BaseTreePruner.java

```diff
@@ -0,0 +1,55 @@
+package org.greenplum.pxf.api.filter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+
+import static org.greenplum.pxf.api.filter.Operator.AND;
+import static org.greenplum.pxf.api.filter.Operator.NOT;
+import static org.greenplum.pxf.api.filter.Operator.OR;
+
+/**
+ * Base tree pruner that maintains tree correctness by analyzing the current node and its children
+ * after both the node and the children have been visited and potentially pruned.
+ */
+public abstract class BaseTreePruner implements TreeVisitor {
+
+    protected final Logger LOG = LoggerFactory.getLogger(this.getClass());
+
+    @Override
+    public Node before(Node node, int level) {
+        return node;
+    }
+
+    @Override
+    public abstract Node visit(Node node, int level);
+
+    @Override
+    public Node after(Node node, int level) {
+        if (node instanceof OperatorNode) {
+            OperatorNode operatorNode = (OperatorNode) node;
+            Operator operator = operatorNode.getOperator();
+            int childCount = operatorNode.childCount();
+            if (AND == operator && childCount == 1) {
+                Node promoted = Optional.ofNullable(operatorNode.getLeft()).orElse(operatorNode.getRight());
+                LOG.debug("Child {} was promoted higher in the tree", promoted);
+                // AND need at least two children. If the operator has a
```

Comment:
I realize I wrote this:
```suggestion
                // AND needs at least two children. If the operator has a
```